#11 Feature templates

amenk opened the pull request:
One Function missing: I want a popup menu entry which I can click, it will detect the current domain and use it as liveDomain and add the environment.
How can I do that?

amenk committed:
#11 Auto configuration of domains by template

amenk committed:
#11 Auto configuration of domains by template: Show Template info always

amenk committed:
#11 Auto configuration of domains: Add entry in the menu to auto-conf environment

amenk commented:
Got it :) Please check and hopefully merge ...

stevehanson commented:
I'm a little busy the next couple days but will try to get a look at this soon. I made some notes on issue #11. Overall, I'm feeling really hesitant to complicate the UI.

amenk committed:
#11 Auto configuration of domains: Bugfix and Styling

amenk committed:
#11 Auto configuration of domains: Styling and texts + Icon

amenk committed:
#11 Auto configuration of domains: Call the add function after document ready

amenk committed:
Code Format

amenk committed:
#15 Improved UI for Templates

- Show template always as last
- Add template not at all projects but just as a button in the bottom

amenk commented:
Improved the UI. Can it maybe be merged now? Would be great!

stevehanson commented:
Overall, it looks pretty good. 

There are a few changes that I'd like to make before pushing it to the store:
- More explanation around the templates feature on the options page, with maybe an example of what a template configuration might look like and information about using with Quick Add.
- Change "Add this" to "Quick Add" on the popup for clarity
- Use of modal instead of JS prompt for capturing the live domain so we can have more information
- Support for arbitrary string in template as opposed to full domains. I think this works, just fails if $$ is used. Not a big deal. Might just need to change terminology, instead of calling this the "live domain", we could call it the template variable or something.
- Live preview of what the generated URLs will be when entering the template string from quick add in the modal
- This is more of a code thing: could we store the template information in $scope.templates (a new array) instead of in $scope.projects. Feel like that might keep the code a little cleaner. I like using an array, because we could in the future support picking which template to use in the quick add modal, and this would keep us from having to migrate users' localstorage data.

If you're not interested in making those changes, I can go ahead and merge this and get to them. I'd like to get these changes implemented before pushing to the webstore.

stevehanson commented:
Would also be nice to have a button on the options page "Add project using template" that launches the modal.

amenk commented:
Thanks for the review!
I think I will try to make the changes - now you make me busy ;)

> This is more of a code thing: could we store the template information in $scope.templates (a new array) instead of in $scope.projects. Feel like that might keep the code a little cleaner. I like using an array, because we could in the future support picking which template to use in the quick add modal, and this would keep us from having to migrate users' localstorage data.

I was having backwards compatibility in mind (not migrating the localstorage). I just can add another array and push it to the local storage as well you mean?

stevehanson commented:
Haha. Yup, your solution was backwards compatible—that's not what I meant. But yeah, creating another array and pushing to local storage is what I was suggesting. 

Lemme know if you want to bounce any questions off me. I'm going to close issue #11. Let's use this PR for discussion instead.

amenk commented:
1. Live domain could be reworded "first entry" (but I think live domain makes also much sense - cause that is usually the plain one, others have added prefixes or suffixes - or do you have another example)
2. But: I will add the quick add also for for example staging domains - then just remove staging and interpolate the live domain (I hope that works)
3. Maybe we do not need the popup at all - I was thinking about live-completing from the template, when the first domain is entered, what do you think?

amenk commented:
Can't come up with a nice UI for alternative 3. (I was thinking about a lock/unlock logic like in graphic software where changing the width affects the height) - will go for a popup as suggested.

amenk commented:
> Haha. Yup, your solution was backwards compatible—that's not what I meant. But yeah, creating another array and pushing to local storage is what I was suggesting.

Currently you store anything in `localStorage['domanSwitcher]` so should I add something like `localStorage['domainSwitcherTemplates']`?

amenk committed:
Templates: Explanations / Placeholders

amenk committed:
Templates: Change "Add this" to "Quick Add" on the popup for clarity

amenk commented:
> Support for arbitrary string in template as opposed to full domains. I think this works, just fails if $$ is used. Not a big deal. Might just need to change terminology, instead of calling this the "live domain", we could call it the template variable or something.

hm .. okay ... the point is that I use the current domain on quick-add ... so it is meant for the live domain - sure you can use any string, but then it is a (only) little bit less awesome :)

amenk committed:
Templates: Creation via modal / drafting / detecting live domain

amenk commented:
Templates are implemented. storage is not changed as I did not get the point :-) (see above).
Can you have a look again and maybe merge?

amenk commented:
_friendly-bump_
Nope, I do not have a calendar reminder for this, even it looks like  this :-)
